### PR TITLE
add packaging API check for non-tanzu cluster

### LINKED ISSUE #340: tanzu package plugin takes too long to respond on non tanzufied clusters
**Bug description**
If you execute `tanzu package available list` on a cluster that does not have kapp-controller 0.20+ installed, it takes between 6 and 10 seconds to get an answer with the following error, followed by the help description:
```
tanzu package available list
\ Retrieving available packages... Error: no matches for kind "PackageMetadata" in version "data.packaging.carvel.dev/v1alpha1"
Usage:
  tanzu package available list or list PACKAGE_NAME [flags]

Examples:

    # List available packages across all namespaces
    tanzu package available list -A

    # List all versions for available package from specified namespace
    tanzu package available list contour.tanzu.vmware.com --namespace test-ns

Flags:
  -h, --help   help for list

Global Flags:
  -A, --all-namespaces      If present, list packages across all namespaces.
      --kubeconfig string   The path to the kubeconfig file, optional
      --log-file string     Log file path
  -n, --namespace string    Namespace of packages, optional (default "default")
  -o, --output string       Output format (yaml|json|table)
      --verbose int32       Number for the log level verbosity(0-9)

Error: exit status 1
```
The error message `Error: no matches for kind "PackageMetadata" in version "data.packaging.carvel.dev/v1alpha1"` is buried in the output, not intuitive for many users and the help message aftwards does not make sense given it's an incompatibility with the cluster and not that the provided CLI command is wrong.

**Affected product area (please put an X in all that apply)**

- [x] APIs
- [ ] Addons
- [ ] CLI
- [ ] Docs
- [ ] Installation
- [X] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Expected behavior**
I would expect the CLI to error in less time, with a more "general user" friendly message (e.g. This cluster does not have Tanzu Package capabilities installed) and without the error message afterwards.

**Steps to reproduce the bug**
- Select a kube context of a non tanzuified cluster or a cluster with kapp-controller less than 0.20
- Execute `tanzu package available list` (time the command)

**Version** (include the SHA if the version is not obvious)
```
tanzu package info
{"name":"package","description":"Tanzu package management","version":"v1.4.0-pre-alpha-2","buildSHA":"5a42a2b","group":"Run","docURL":"","completionType":0}
```

**Environment where the bug was observed (cloud, OS, etc)**
macOS

**Relevant Debug Output (Logs, manifests, etc)**
Since there's no --verbose output for this plugin or --log-file, nothing can be provided.

### PR DESCRIPTION
**Pull request by ggpaue**
### What this PR does / why we need it
If cluster doesn't have Carvel packaging API available, package plugin will return error immediately

### Which issue(s) this PR fixes

Fixes #340 

### Describe testing done for PR

Passed package plugin integration test and manually tested with a cluster without Carvel packaging api-resources

### Release note

```release-note
check Carvel packaging API before package plugin cmd executes
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: cmd/cli/plugin/package/main.go [hunk 1]
```diff
@@ -52,3 +54,48 @@ func getOutputFormat() string {
...
+	if err != nil {
+		return false, err
+	}
+	clusterQueryClient, err := capdiscovery.NewClusterQueryClientForConfig(cfg)
+	if err != nil {
+		log.Error(err, "failed to create a new instance of the cluster query builder")
+		return false, err
+	}
+
+	apiGroup := capdiscovery.Group("packageMetadateAPIQuery", "data.packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packagemetadatas")
```
Comment (vijaykatam) [inline cmd/cli/plugin/package/main.go:69]: I think you can combine all three queries into one with discovery client. cc @rajathagasthya

Comment (rajathagasthya) [reply cmd/cli/plugin/package/main.go:69]: Yep, you can combine by adding multiple arguments to `Query()`.

```go
q1 := capdiscovery.Group("packageMetadateAPIQuery", "data.packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packagemetadatas")
q2 := capdiscovery.Group("packageRepositoryAPIQuery", "packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packagerepositories")
q3 := capdiscovery.Group("packageInstallAPIQuery", "packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packageinstalls")

clusterQueryClient.Query(q1, q2, q3).Execute()
```

File: cmd/cli/plugin/package/package_install.go [hunk 2]
```diff
@@ -25,8 +25,9 @@ var packageInstallCmd = &cobra.Command{
 	
     # Install package contour with kubeconfig flag and waiting for package reconciliation to complete	
     tanzu package install contour-pkg --package-name contour.tanzu.vmware.com --namespace test-ns --version 1.15.1-tkg.1-vmware1 --kubeconfig path/to/kubeconfig`,
-	RunE:         packageInstall,
-	SilenceUsage: true,
+	RunE:              packageInstall,
+	SilenceUsage:      true,
+	PersistentPreRunE: packageMetadataAvailabilityCheck,
```
Comment (maralavi) [inline cmd/cli/plugin/package/package_install.go:30]: I think ```"package install"``` and ```"package installed"``` operation should use the same availability call (they need both ```"packagemetadatas"``` and ```"packageinstalls"``` api to be available). But right now, ```"package install"``` is using ```"packageMetadataAvailabilityCheck"``` and ```"package installed"``` operations are using ```"packageInstallAvailabilityCheck"```.

File: cmd/cli/plugin/package/main.go (hunk 1 above)
Comment (ggpaue) [reply cmd/cli/plugin/package/main.go:69]: thanks for the review @vijaykatam , I'm wondering should we combine all the API checks together?

for example, `package available` only needs packageMetadateAPI check, `package repository` only needs packageRepositoryAPI check

Comment (ggpaue) [reply cmd/cli/plugin/package/main.go:69]: bring @maralavi into this discussion

Comment (maralavi) [reply cmd/cli/plugin/package/main.go:69]: @ggpaue It won't hurt to check the availability of all apis for all commands. Anyways whoever is going to use package plugin needs all of those be available (eventhough when one of those available, almost always the remaining related apis for packaging are also available). I assume the discovery calls are quite fast and you could merge all these 3 into one query and use this combined approach as suggested above, and you could call that combined query for all commands.

File: cmd/cli/plugin/package/package_install.go (hunk 2 above)
Comment (ggpaue) [reply cmd/cli/plugin/package/package_install.go]: thanks for the review @maralavi , now we put all the checks together

File: cmd/cli/plugin/package/main.go (hunk 1 above)
Comment (ggpaue) [reply cmd/cli/plugin/package/main.go]: updated the code to put all API queries together

File: cmd/cli/plugin/package/main.go [hunk 3]
```diff
@@ -52,3 +54,22 @@ func getOutputFormat() string {
 	}
 	return format
 }
+
+func isCarvelAPIAvailable(kubeCfgPath string) (bool, error) {
```
Comment (maralavi) [inline cmd/cli/plugin/package/main.go:58]: Isn't ```isPackagingAPIAvailable``` a slightly better naming?

File: cmd/cli/plugin/package/package_available.go [hunk 4]
```diff
@@ -12,14 +15,27 @@ import (
...
+	PersistentPreRunE: carvelAvailabilityCheck,
 }
 
 func init() {
 	packageAvailableCmd.PersistentFlags().StringVarP(&packageAvailableOp.KubeConfig, "kubeconfig", "", "", "The path to the kubeconfig file, optional")
 	packageAvailableCmd.PersistentFlags().StringVarP(&packageAvailableOp.Namespace, "namespace", "n", "default", "Namespace of packages, optional")
 	packageAvailableCmd.PersistentFlags().StringVarP(&outputFormat, "output", "o", "", "Output format (yaml|json|table), optional")
 }
+
+func carvelAvailabilityCheck(_ *cobra.Command, _ []string) error {
```
Comment (maralavi) [inline cmd/cli/plugin/package/package_available.go:31]: how about ```packagingAPIAvailabilityCheck```?

File: cmd/cli/plugin/package/main.go (hunk 3 above)
Comment (ggpaue) [reply cmd/cli/plugin/package/main.go]: thanks for the naming advice @maralavi 
updated the function name accordingly

**Review by maralavi — APPROVED** [review]

**Comment by danniel1205** [comment]
Questions:
- Shall we update the release notes in description ? It looks like belongs to `Tests done` section. 
- Shall we add unit/integ tests to cover the newly added function, especially for negative cases ?

File: cmd/cli/plugin/package/main.go [hunk 5]
```diff
@@ -52,3 +54,22 @@ func getOutputFormat() string {
...
+	clusterQueryClient, err := capdiscovery.NewClusterQueryClientForConfig(cfg)
+	if err != nil {
+		log.Error(err, "failed to create a new instance of the cluster query builder")
+		return false, err
+	}
+
+	apiGroup1 := capdiscovery.Group("packageMetadateAPIQuery", "data.packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packagemetadatas")
+	apiGroup2 := capdiscovery.Group("packageAPIQuery", "data.packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packages")
+	apiGroup3 := capdiscovery.Group("packageRepositoryAPIQuery", "packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packagerepositories")
+	apiGroup4 := capdiscovery.Group("packageInstallAPIQuery", "packaging.carvel.dev").WithVersions("v1alpha1").WithResource("packageinstalls")
```
Comment (danniel1205) [inline cmd/cli/plugin/package/main.go:72]: question: Is there a way we could discover resources without hardcoding the version? A use case I could think of is:

* I have v1alpha1 carvel with my current cluster
* I upgrade my cluster which has v1beta1 carvel now
* My current package plugin will break because there will be no `v1alpha1` anymore

Comment (ggpaue) [reply cmd/cli/plugin/package/main.go:72]: not sure how to discover the API version but I believe the upgrade case will also affect secretGen API version

https://github.com/vmware-tanzu/tanzu-framework/blob/main/cmd/cli/plugin/secret/main.go

cc @maralavi

Comment (danniel1205) [reply cmd/cli/plugin/package/main.go:72]: By looking at this [code](https://github.com/vmware-tanzu/tanzu-framework/blob/978797e5abd67ee1a8b0e1a087340f87e4ae3dd4/pkg/v1/sdk/capabilities/discovery/cluster_gvr.go#L51), I dont think `WithVersions` is required. Could you help test it out ?

**Comment by danniel1205** [comment]
Adding `do-not-merge/hold` because this [thread](https://github.com/vmware-tanzu/tanzu-framework/pull/1242#discussion_r761528548) needs to be addressed.

**Comment by ggpaue** [comment]
> Questions:
> 
> * Shall we update the release notes in description ? It looks like belongs to `Tests done` section.
> * Shall we add unit/integ tests to cover the newly added function, especially for negative cases ?

thanks @danniel1205 , the packaging API check is in each command `PersistentPreRunE`, so the current integration test will cover this check. I will add another case which will delete kapp then see if it can return error

File: cmd/cli/plugin/package/main.go (hunk 5 above)
Comment (ggpaue) [reply cmd/cli/plugin/package/main.go:72]: yeah I've deleted `WithVersions` and it will fall into this error:
https://github.com/vmware-tanzu/tanzu-framework/blob/978797e5abd67ee1a8b0e1a087340f87e4ae3dd4/pkg/v1/sdk/capabilities/discovery/cluster_gvr.go#L53
`
Error: failed to check for the availability of 'packaging.carvel.dev/v1alpha1' API: cannot check for group and resource existence without version info; use WithVersion method`

Comment (ggpaue) [reply cmd/cli/plugin/package/main.go]: added constants to represent API name and version instead of hardcoding

**Review by danniel1205 — APPROVED** [review]
LGTM

**Comment by danniel1205** [comment]
A github [issue](https://github.com/vmware-tanzu/tanzu-framework/issues/1294) has been filed to follow up on this discussion: https://github.com/vmware-tanzu/tanzu-framework/pull/1242#discussion_r761528548. We might want to switch to new discover API once it allows us to discover resources without specifying a version.

**Review by blc1996 — APPROVED** [review]
LGTM
